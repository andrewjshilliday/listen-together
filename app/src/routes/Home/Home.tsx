import React, { useState, useEffect } from 'react';
import { useHistory } from "react-router-dom";
import { Button, TextField, InputAdornment, IconButton } from '@material-ui/core';
import ArrowForwardIcon from '@material-ui/icons/ArrowForward';
import { useRoom, useAuthorization } from '../../components/providers';
import styled from 'styled-components';
import { CSSTransition } from 'react-transition-group';
import musicSvg from '../../assets/images/listen-together.svg';

const Home: React.FC = (props: any) => {
  const [name, setName] = useState('');
  const [roomId, setRoomId] = useState('');
  const history = useHistory();
  const roomProvider = useRoom();
  const authProvider = useAuthorization();

  useEffect(() => {
    if (roomProvider.roomId) {
      props.history.push(`/room/${roomProvider.roomId}`);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const createUser = (userName: string) => {
    roomProvider.actions.createUser(userName);
  }

  const createRoom = () => {
    roomProvider.actions.createRoom((roomId: string) => {
      if (roomId) {
        history.push(`/room/${roomId}`);
      }
    });
  };

  const joinRoom = () => {
    roomProvider.actions.joinRoom(roomId, (roomId: string) => {
      if (roomId) {
        history.push(`/room/${roomId}`);
      }
    });
  }

  return (
    <HomeContainer>
      <CSSTransition in={!authProvider.authorized} classNames="fade" timeout={300} unmountOnExit>
        <FadeInContainer>
          <Button variant="contained" color="primary" onClick={() => authProvider.actions.signIn()}>Sign In</Button>
        </FadeInContainer>
      </CSSTransition>
      <CSSTransition in={authProvider.authorized} classNames="fade" timeout={300} unmountOnExit>
        <FadeInContainer>
          <Title>ListenTogether</Title>
          <ListenTogetherImage src={musicSvg} />
          <UsernameContainer disabled={roomProvider.username != null}>
            <h2>First, Pick a Username</h2>
            <UsernameInput label="Enter Username" onChange={(e) => setName(e.target.value)}
              onKeyPress={(e) => {
                if (e.key === 'Enter') {
                  e.preventDefault();
                  createUser(name);
                }
              }}
              InputProps={{ endAdornment: (
                <InputAdornment position="end">
                  <IconButton onClick={() => createUser(name)}>
                    <ArrowForwardIcon />
                  </IconButton>
                </InputAdornment>
              ) }}
            />
          </UsernameContainer>
          <RoomChoicesContainer disabled={roomProvider.username == null}>
            <h2>Then</h2>
            <CreateButton variant="contained" color="primary" onClick={() => createRoom()}>Create a Room</CreateButton>
            <h3>Or</h3>
            <JoinRoomContainer>
              <JoinInput label="Room ID" onChange={(e) => setRoomId(e.target.value)} />
              <JoinButton variant="contained" color="primary" onClick={() => joinRoom()}>Join a Room</JoinButton>
            </JoinRoomContainer>
          </RoomChoicesContainer>
        </FadeInContainer>
      </CSSTransition>
    </HomeContainer>
  );
}

export default Home;


const HomeContainer = styled.div`
  display: flex;
  flex-direction: column;
  text-align: center;
  height: 100vh;
  background: #99ccff linear-gradient(315deg, #99ccff, #e6f2ff);
  background-size: 100% 100%;
  overflow: auto;

  h1, h2, h3 { margin-bottom: 10px; }
`;
const FadeInContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: auto;
  transition: opacity 300ms ease-in;
  &.fade-enter { opacity: 0; }
  &.fade-enter-active { opacity: 1; }
  &.fade-exit { opacity: 1; }
  &.fade-exit-active { opacity: 0; }
`;
const Title = styled.h1`
  font-size: 4em;
`;
const ListenTogetherImage = styled.img`
  height: 200px;
  width: 300px;
  margin-bottom: 20px;
`;
const UsernameContainer = styled.div<{disabled: boolean}>`
  opacity: 1;
  transition: opacity 500ms ease-in-out;
  ${props => props.disabled && `
    pointer-events: none;
    opacity: 0.1;
  `}
`;
const UsernameInput = styled(TextField)`
  height: 50px;
  width: 300px;
  margin-bottom: 20px !important;
`;
const RoomChoicesContainer = styled.div<{disabled: boolean}>`
  opacity: 1;
  transition: opacity 500ms ease-in-out;
  ${props => props.disabled && `
    pointer-events: none;
    opacity: 0.1;
  `}
`;
const CreateButton = styled(Button)`
  height: 50px;
  width: 300px;
  margin-bottom: 20px !important;
`;
const JoinRoomContainer = styled.div`
  display: flex;
  height: 50px;
  margin-bottom: 20px;
`;
const JoinInput = styled(TextField)`
  width: 150px;  
`;
const JoinButton = styled(Button)`
  width: 150px;
`;
