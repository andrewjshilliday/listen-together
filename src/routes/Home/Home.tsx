import React from 'react';
import { useHistory } from "react-router-dom";
import Button from '@material-ui/core/Button';
import { useRoom } from '../../components/providers';
import './Home.scss';

const Home: React.FC = () => {
  const history = useHistory();
  const room = useRoom();

  const createRoom = () => {
    const roomId = 'DJGFE';
    if (!room.roomId) {
      room.setRoomId(roomId);
    }
    history.push(`/room/${roomId}`);
  };

  return (
    <div className="home view">
      <Button className="create-button" variant="contained" color="primary" onClick={() => createRoom()}>Create a Room</Button>
    </div>
  );
}

export default Home;
