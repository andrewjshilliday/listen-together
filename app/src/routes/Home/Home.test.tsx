import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import renderer, { act } from 'react-test-renderer';
import Home from './Home';

jest.mock('../../components/providers/AuthorizationProvider', () => require('../../__mocks__/providers/AuthorizationProvider.mock').default);
jest.mock('../../components/providers/RoomProvider', () => require('../../__mocks__/providers/RoomProvider.mock').notLoggedIn);

jest.useFakeTimers();

beforeAll(() => {
  jest.resetModules();
})

it('renders correctly', async () => {
  const tree = renderer.create(
    <MemoryRouter initialEntries={[ '/home' ]}>
      <Route path="/home" component={Home} />
    </MemoryRouter>);

  jest.runTimersToTime(1);

  await act(async () => {
    await Promise.resolve();
  });

  expect(tree.toJSON()).toMatchSnapshot();
});
