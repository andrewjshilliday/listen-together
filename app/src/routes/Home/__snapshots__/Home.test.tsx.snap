// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-AxjAm bcStPz"
>
  <h1
    className="sc-AxiKw OViUk"
  >
    ListenTogether
  </h1>
  <img
    className="sc-AxhCb ivzhVo"
    src="listen-together.svg"
  />
  <div
    className="sc-AxirZ bGbxqr"
  >
    <div
      className="sc-AxhUy jOsxxC"
      disabled={false}
    >
      <h2>
        First, Pick a Username
      </h2>
      <div
        className="MuiFormControl-root MuiTextField-root sc-AxgMl kiQBjK"
        onKeyPress={[Function]}
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
        >
          Enter Username
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
          />
          <div
            className="MuiInputAdornment-root MuiInputAdornment-positionEnd"
          >
            <button
              className="MuiButtonBase-root MuiIconButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
                  />
                </svg>
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-AxheI cZzwDd"
      disabled={true}
    >
      <h2>
        Then
      </h2>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained sc-Axmtr fXfIRP MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Create a Room
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
      <h3>
        Or
      </h3>
      <div
        className="sc-AxmLO klesxR"
      >
        <div
          className="MuiFormControl-root MuiTextField-root sc-fzozJi bhOwZD"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
          >
            Room ID
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
            />
          </div>
        </div>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained sc-fzoLsD eYvSSR MuiButton-containedPrimary"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Join a Room
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
